### PR TITLE
Functions refactor

### PR DESCRIPTION
**Pull request by cpcallen**
Create several new subclasses of `Interpreter.prototype.Function`:
* `UserFunction` encapsulates creating and calling user-defined (JS) functions.
* `NativeFunction` encapsulates native functions (i.e., ones implemented in or by node.js).
   * `OldNativeFunction` is a subclass for the existing native functions created using `createNativeFunction`.
   * `OldAsyncFunction` is a subclass for the existing native functions created using `createAsyncFunction`.

The intention is ultimately to rewrite all native functions to use the new API provided by `NativeFunction`, then remove the `Old` classes as well as their corresponding `create` functions.  For now, some key examples have been converted.

### CODE REVIEW

File: server/interpreter.js [hunk 1]
```diff
@@ -3043,7 +3376,7 @@ Interpreter.prototype.installTypes = function() {
   };
 
   /**
-   * Server is an (owner, port, proto, (extra info)) tuple representing a
+   * Server is an (owner, port, proto, (extra info) tuple representing a
```
Comment (NeilFraser) [inline server/interpreter.js]: I think that ) was needed.

File: server/interpreter.js [hunk 2]
```diff
@@ -2797,26 +2818,346 @@ Interpreter.prototype.installTypes = function() {
...
+    if (!node) { // Deserializing
+      this.node = /** @type {?} */ (null);
+      this.scope = /** @type {?} */ (null);
+      this.source = '';
+      return;
     }
+    intrp.Function.call(/** @type {?} */ (this), owner, proto);
+    this.node = node;
+    this.scope = scope;
+    if(node['id']) {
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/interpreter.js]: Fixed here and elsewhere.

By the way: we are still inconsistent about `function()` vs `function ()`.  The style guide mandates the latter, though the former appears to be very popular too.  We do both for good measure…

File: server/interpreter.js [hunk 3]
```diff
@@ -2797,26 +2818,346 @@ Interpreter.prototype.installTypes = function() {
...
+    this.source = src.substring(node['start'], node['end']);
+    // Record the (whole) source on the function node's body node, for
+    // use by the the (pseudo)Error constructor when generating stack
+    // traces (via Thread.prototype.getSource).  We store it on the
+    // body node (rather than on the function node) because the
+    // function node never appears on the stateStack_ when the
+    // function is being executed.  We save the full original source
+    // (not just the part containing the function) because the start
+    // and end offsets on the AST nodes are absolute, not relative.
+    if (!(node['body']['source'])) {
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/interpreter.js]: Fixed.

File: server/interpreter.js [hunk 4]
```diff
@@ -2797,26 +2818,346 @@ Interpreter.prototype.installTypes = function() {
...
+    }
+    intrp.setProperty(this, 'name', name, Descriptor.c);
   };
 
+  /**
+   * The [[Call]] internal method defined by §13.2.1 of the ES5.1 spec.
+   * @override
+   */
+  intrp.UserFunction.prototype.call = function(
+      intrp, thread, state, thisVal, args) {
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/interpreter.js]: There is a JSDoc.  You mean no annotation of the params?  See 
9899084 and https://google.github.io/styleguide/jsguide.html#features-functions-parameters (internal styleguide agrees with external).

File: server/interpreter.js [hunk 5]
```diff
@@ -2797,26 +2818,346 @@ Interpreter.prototype.installTypes = function() {
...
+    // created, and (2) functions created using the Function
+    // constructor have this.scope set to the global scope, which is
+    // owned by root!
+    var scope = new Interpreter.Scope(this.owner, this.scope);
+    intrp.populateScope_(this.node['body'], scope, this.source);
+    // Add all arguments.
+    for (var i = 0; i < this.node['params'].length; i++) {
+      var paramName = this.node['params'][i]['name'];
+      var paramValue = args.length > i ? args[i] :
+          undefined;
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/interpreter.js]: So it does!

File: server/interpreter.js [hunk 6]
```diff
@@ -2797,26 +2818,346 @@ Interpreter.prototype.installTypes = function() {
...
+    // this.scope.perms, for the new scope perms because (1) we want
+    // to be able to change the owner of a function after it's
+    // created, and (2) functions created using the Function
+    // constructor have this.scope set to the global scope, which is
+    // owned by root!
+    var scope = new Interpreter.Scope(this.owner, this.scope);
+    intrp.populateScope_(this.node['body'], scope, this.source);
+    // Add all arguments.
+    for (var i = 0; i < this.node['params'].length; i++) {
+      var paramName = this.node['params'][i]['name'];
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/interpreter.js]: Probably not actually faster.  But easier to read.

File: server/interpreter.js [hunk 7]
```diff
@@ -2797,26 +2818,346 @@ Interpreter.prototype.installTypes = function() {
...
+  intrp.UserFunction.prototype.construct = function(
+      intrp, thread, state, args) {
+    if (!state.object_) {
+      var proto = intrp.getProperty(this, 'prototype');
+      // Per ES5.1 §13.2.2 step 7: if .prototype is primitive, use
+      // Object.prototype instead.
+      if (!(proto instanceof intrp.Object)) {
+        proto = intrp.OBJECT;
+      }
+      state.object_= new intrp.Object(state.scope.perms, proto);
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/interpreter.js]: Fixed.

File: server/interpreter.js (hunk 1 above)
Comment (cpcallen) [reply server/interpreter.js]: Indeed.  Not sure how that happened; fat fingers, probably!

**Comment by cpcallen** [comment]
I neglected to sync my branch to GitHub before doing the rebase-and-merge, so My fixes for this PR were inadvertently omitted from the final commit.  Pushing them separately (and without a further PR).

Maybe we should have a policy of never doing the final commit for our own PRs, so that the reviewer can catch errors like this.

**Pull request merged**